test/microbench/org/apache/cassandra/test/microbench/sstable/SSTableReaderBench.java

```diff
@@ -0,0 +1,209 @@
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.TearDown;
+import org.openjdk.jmh.annotations.Threads;
+import org.openjdk.jmh.annotations.Warmup;
+
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Warmup(iterations = 1, time = 5)
+@Measurement(iterations = 3, time = 5)
+@Fork(value = 1, jvmArgsAppend = "-Xmx4G")
+@Threads(1)
+@State(Scope.Benchmark)
+public class SSTableReaderBench extends AbstractSSTableBench
+{
+    private final static Logger logger = LoggerFactory.getLogger(SSTableReaderBench.class);
+
+    int KEY_SIZE = 8;
+    int P_KEYS = 10000;
+    int C_KEYS = 1000;
+    int VAL_SIZE = 1;
+
+    public ByteBuffer[] ckeys;
+    public DecoratedKey[] pkeys;
+    public DecoratedKey[] nonpkeys;
+
+    private SSTableReader sstr;
+
+    private int idx = 0;
+
+    private final int step = P_KEYS / 2 - 1;
+
+    @Param({ "table", "table_with_clustering" })
+    public String tableName;
+
+    @Param({ "BIG", "BTI" })
+    public String formatName;
+    private ColumnFamilyStore table;
+    private LifecycleTransaction txn;
+
+    @Setup
+    public void setup()
+    {
+        Keyspace ks = prepareMetadata();
+        table = ks.getColumnFamilyStore(tableName);
+        pkeys = prepareDecoratedKeys(0, P_KEYS, KEY_SIZE);
+        nonpkeys = prepareDecoratedKeys(P_KEYS, P_KEYS * 2, KEY_SIZE);
+        ckeys = prepareBuffers(0, C_KEYS, KEY_SIZE);
+    }
+
+    @Setup(Level.Iteration)
+    public void setupIteration() throws Exception
+    {
+        txn = LifecycleTransaction.offline(OperationType.WRITE);
+        sstr = prepareReader(getFormat(formatName), table, txn);
+    }
+
+    private int nextIdx() {
+        idx += step;
+        if (idx >= P_KEYS)
+            idx -= P_KEYS;
+        return idx;
+    }
+
+    @Benchmark
+    @BenchmarkMode(Mode.Throughput)
+    public void getEQPosition()
+    {
+        sstr.getPosition(pkeys[nextIdx()], SSTableReader.Operator.EQ);
+    }
+
+    @Benchmark
+    @BenchmarkMode(Mode.Throughput)
+    public void getEQPositionNonExisting()
+    {
+        sstr.getPosition(nonpkeys[nextIdx()], SSTableReader.Operator.EQ);
+    }
+
+    @Benchmark
+    @BenchmarkMode(Mode.Throughput)
+    public void getGTPosition()
+    {
+        sstr.getPosition(pkeys[nextIdx()], SSTableReader.Operator.GT);
+    }
+
+    @Benchmark
+    @BenchmarkMode(Mode.Throughput)
+    public void getGTPositionNonExisting()
+    {
+        sstr.getPosition(nonpkeys[nextIdx()], SSTableReader.Operator.GT);
+    }
+
+    @Benchmark
+    @BenchmarkMode(Mode.AverageTime)
+    public void iterateOverAllKeys() throws Exception
+    {
+        try (PartitionIndexIterator it = sstr.allKeysIterator())
+        {
+            while (!it.isExhausted()) it.advance();
```

Comment:
I hope it will not get optimized...

Reply:
why would it be? this is a side-effect operation